escrow/memory_storage.go

```diff
@@ -110,13 +109,104 @@ func (storage *memoryStorage) Clear() (err error) {
 }
 
 func (storage *memoryStorage) StartTransaction(conditionKeys []string) (transaction Transaction, err error) {
-	return nil, fmt.Errorf("Not implemented")
+	conditionKeyValues := make([]KeyValueData, len(conditionKeys))
+	for i, key := range conditionKeys {
+		value, ok, err := storage.Get(key)
+		if err != nil {
+			return nil, err
+		} else if !ok {
+			conditionKeyValues[i] = KeyValueData{Key: key, Value: "", Present: false}
+		} else {
+			conditionKeyValues[i] = KeyValueData{Key: key, Value: value, Present: true}
+		}
+
+	}
+	transaction = &memoryStorageTransaction{ConditionKeys: conditionKeys, ConditionValues: conditionKeyValues}
+	return transaction, nil
 }
 
+func getValueDataForKey(key string, update []KeyValueData) (data KeyValueData, present bool) {
+	for _, data := range update {
+		if strings.Compare(data.Key, key) == 0 {
+			return data, true
+		}
+	}
+	return data, false
+}
 func (storage *memoryStorage) CompleteTransaction(transaction Transaction, update []KeyValueData) (ok bool, err error) {
-	return false, fmt.Errorf("Not implemented")
+	originalValues := transaction.(*memoryStorageTransaction).ConditionValues
+	for _, olddata := range originalValues {
+		if olddata.Present {
+			//make sure the current value is the same as the value last read
+			currentValue, ok, err := storage.Get(olddata.Key)
+			if !ok || err != nil {
+				return ok, err
+			}
+			if strings.Compare(currentValue, olddata.Value) == 0 {
+				if updatedData, ok := getValueDataForKey(olddata.Key, update); ok {
+					if err = storage.Put(updatedData.Key, updatedData.Value); err != nil {
+						return false, err
+					}
+					continue
+				}
+			}
+
+		} else {
+			if updatedData, ok := getValueDataForKey(olddata.Key, update); ok {
+				if ok, err := storage.PutIfAbsent(updatedData.Key, updatedData.Value); err != nil {
+					return false, err
+				} else if !ok {
+					return ok, nil
+				}
+				continue
+			}
+		}
+	}
+	return true, nil
 }
 
 func (client *memoryStorage) ExecuteTransaction(request CASRequest) (ok bool, err error) {
-	return false, fmt.Errorf("Not implemented")
+
+	transaction, err := client.StartTransaction(request.ConditionKeys)
```

Comment:
You can do it in separate PR as well, but `ExecuteTransaction` method implementation also could be reused for all kind of storages.

Reply:
Sure, let me create a issue on this and will raise a seprate PR , #507